### PR TITLE
Reconcile CABundle in webhook configurations

### PR DESCRIPTION
**Pull request by mfaizanse**
**Description**

Changes proposed in this pull request:

- Reconcile CABundle in webhook configurations

**Related issue(s)**

- https://github.com/kyma-project/eventing-manager/issues/51

### CODE REVIEW

File: test/utils/integration/integration.go [hunk 1]
```diff
@@ -873,6 +941,37 @@ func (env TestEnvironment) GetServiceFromK8s(name, namespace string) (*corev1.Se
...
+	nn := types.NamespacedName{
+		Name:      name,
+		Namespace: namespace,
+	}
+	result := &corev1.Secret{}
+	if err := env.k8sClient.Get(env.Context, nn, result); err != nil {
+		return nil, err
+	}
+	return result, nil
+}
```
Comment (muralov) [inline test/utils/integration/integration.go:954]: nitpick: maybe create a common `GetSecret` function in KubeClient, which can be used by `func (r *Reconciler) reconcileWebhooksWithCABundle(ctx context.Context) error` in weebhook.go

File: test/utils/integration/integration.go [hunk 2]
```diff
@@ -873,6 +941,37 @@ func (env TestEnvironment) GetServiceFromK8s(name, namespace string) (*corev1.Se
...
+	}
+	var validatingWH admissionv1.ValidatingWebhookConfiguration
+	validatingWHKey := client.ObjectKey{
+		Name: getTestBackendConfig().ValidatingWebhookName,
+	}
+	if err := env.k8sClient.Get(ctx, validatingWHKey, &validatingWH); err != nil {
+		return nil, nil, err
+	}
+	return &mutatingWH, &validatingWH, nil
+}
```
Comment (muralov) [inline test/utils/integration/integration.go:973]: this one seems to me duplicate as you have this `func (r *Reconciler) getMutatingAndValidatingWebHookConfig(ctx context.Context)` in webhook.go file too. Let's create a common function in KubeClient?

File: test/utils/integration/integration.go [hunk 3]
```diff
@@ -834,9 +901,10 @@ func getValidatingWebhookConfig(webhook []admissionv1.ValidatingWebhook) *admiss
 
 func getTestBackendConfig() env.BackendConfig {
 	return env.BackendConfig{
-		WebhookSecretName:     "webhookSecret",
-		MutatingWebhookName:   "mutatingWH",
-		ValidatingWebhookName: "validatingWH",
+		WebhookSecretName:     "eventing-manager-webhook-server-cert",
+		MutatingWebhookName:   "subscription-mutating-webhook-configuration",
+		ValidatingWebhookName: "subscription-validating-webhook-configuration",
+		Namespace:             "kyma-system",
```
[the remark this answers is not in the document]
Comment (mfaizanse) [reply test/utils/integration/integration.go:907]: The webhook secret is created at the start of envTest and its shared/used all test cases. That's why its only created in `kyma-system` namespace. We do not create it per test case.

File: internal/controller/eventing/webhook_test.go [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+					{
+						ClientConfig: admissionv1.WebhookClientConfig{},
+					},
+				}),
+				getValidatingWebhookConfig(nil),
+			},
+			wantError: errInvalidObject,
+		},
+		{
+			name: "WH does not contain valid CABundle",
```
[the remark this answers is not in the document]
Comment (mfaizanse) [reply internal/controller/eventing/webhook_test.go]: updated!

File: internal/controller/eventing/webhook_test.go [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+				{
+					ClientConfig: admissionv1.WebhookClientConfig{
+						CABundle: dummyCABundle,
+					},
+				},
+			}),
+			wantError: nil,
+		},
+		{
+			name: "WH contains valid CABundle",
```
[the remark this answers is not in the document]
Comment (mfaizanse) [reply internal/controller/eventing/webhook_test.go]: updated!

File: internal/controller/eventing/webhook_test.go [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+				{
+					ClientConfig: admissionv1.WebhookClientConfig{
+						CABundle: dummyCABundle,
+					},
+				},
+			}),
+			wantError: nil,
+		},
+		{
+			name: "WH contains outdated valid CABundle",
```
[the remark this answers is not in the document]
Comment (mfaizanse) [reply internal/controller/eventing/webhook_test.go]: updated!

File: test/utils/integration/integration.go [hunk 7]
```diff
@@ -675,6 +691,41 @@ func (env TestEnvironment) EnsureEPPClusterRoleBindingCorrect(t *testing.T, even
...
+		if !bytes.Equal(mwh.Webhooks[0].ClientConfig.CABundle, certSecret.Data[eventingctrl.TLSCertField]) {
+			env.Logger.WithContext().Error("CABundle of mutating configuration is not correct")
+			return false
+		}
+
+		if !bytes.Equal(vwh.Webhooks[0].ClientConfig.CABundle, certSecret.Data[eventingctrl.TLSCertField]) {
+			env.Logger.WithContext().Error("CABundle of validating configuration is not correct")
+			return false
+		}
+		return true
```
[the remark this answers is not in the document]
Comment (mfaizanse) [reply test/utils/integration/integration.go]: This code is intended to test the logic of `func (r *Reconciler) reconcileWebhooksWithCABundle(ctx context.Context) error` thats why it seems similar.

File: test/utils/integration/integration.go (hunk 1 above)
Comment (mfaizanse) [reply test/utils/integration/integration.go]: updated it to use the common method from `KubeClient`.

File: test/utils/integration/integration.go (hunk 2 above)
Comment (mfaizanse) [reply test/utils/integration/integration.go]: updated it to use the common methods from `KubeClient`.

File: internal/controller/eventing/controller.go [hunk 8]
```diff
@@ -204,6 +209,13 @@ func (r *Reconciler) handleEventingReconcile(ctx context.Context,
 	// set state processing if not set yet
 	r.InitStateProcessing(eventing)
 
+	// sync webhooks CABundle.
+	if err := r.reconcileWebhooksWithCABundle(ctx); err != nil {
+		return ctrl.Result{}, r.syncStatusWithWebhookErr(ctx, eventing, err, log)
+	}
+	// set webhook condition to true.
+	eventing.Status.SetWebhookReadyConditionToTrue()
```
[the remark this answers is not in the document]
Comment (mfaizanse) [reply internal/controller/eventing/controller.go:217]: IMO, yes if something is wrong with webhooks then we can reflect the information in the Eventing CR status here.

[the remark this answers is not in the document]
Comment (muralov) [reply internal/controller/eventing/controller.go:217]: It is okay, we can have it. Actually, this should not part of the business logic as attaching caBundle is  certificate-managers job. We don't have it and we'll anyway get rid of this hack in the future when we get rid of v1alpha1. Consequently, the condition too.

**Review by muralov — APPROVED** [review]

**Pull request merged**